src/java.base/share/classes/sun/security/pkcs/NamedPKCS8Key.java

```diff
@@ -0,0 +1,113 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package sun.security.pkcs;
+
+import sun.security.util.DerInputStream;
+import sun.security.util.DerValue;
+import sun.security.x509.AlgorithmId;
+
+import java.io.IOException;
+import java.io.InvalidObjectException;
+import java.io.ObjectInputStream;
+import java.io.Serial;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.ProviderException;
+import java.security.spec.NamedParameterSpec;
+
+public final class NamedPKCS8Key extends PKCS8Key {
+    @Serial
+    private static final long serialVersionUID = 1L;
+
+    private final String fname;
+    private final transient NamedParameterSpec paramSpec;
+    private final byte[] h;
+
+    // Ctor from family name, parameter set name, raw key bytes.
+    // Key bytes won't be cloned, caller must relinquish ownership
+    public NamedPKCS8Key(String fname, String pname, byte[] h) {
+        this.fname = fname;
+        this.paramSpec = new NamedParameterSpec(pname);
+        try {
+            this.algid = AlgorithmId.get(pname);
+        } catch (NoSuchAlgorithmException e) {
+            throw new ProviderException(e);
+        }
+        this.h = h;
+
+        DerValue val = new DerValue(DerValue.tag_OctetString, h);
+        try {
+            this.key = val.toByteArray();
+        } finally {
+            val.clear();
+        }
+    }
+
+    // Ctor from family name, and PKCS #8 bytes
+    public NamedPKCS8Key(String fname, byte[] encoded) throws InvalidKeyException {
+        super(encoded);
+        this.fname = fname;
+        try {
+            paramSpec = new NamedParameterSpec(algid.getName());
+            if (algid.getEncodedParams() != null) {
+                throw new InvalidKeyException("algorithm identifier has params");
+            }
+            h = new DerInputStream(key).getOctetString();
+        } catch (IOException e) {
+            throw new InvalidKeyException("Cannot parse input", e);
+        }
+    }
+
+    @Override
+    public String toString() {
+        // Do not modify: this can be used by earlier JDKs that
+        // does not have the getParams() method
+        return paramSpec.getName() + " private key";
+    }
+
+    // Returns the reference to the internal key. Caller must not modify
+    // the content or keep a reference
+    public byte[] getRawBytes() {
+        return h;
```

Comment:
Should there be a defensive copy here?  I realize it is within sun.security.pkcs so maybe it's less of an issue to hand the mutable reference back, but I would consider the defensive copy if it won't severly impact performance.

Reply:
See the comments at the beginning of `NamedKeyPairGenerator`. I try my best to get the ownership of byte arrays to be as clear as possible to avoid unnecessary duplicates. Do you still see a problem here?